grammar/test/test_grammar.py

```diff
@@ -122,92 +133,42 @@ def test_comma_var():
-
-def test_galaxy():
-    '''
-    hot_zone = randint(1, 8)
-    bio_zone = randint(hot_zone + 1, hot_zone + 15)
-    cold_zone = randint(bio_zone + 1, bio_zone + 20)
-    gas_zone = randint(cold_zone + 1, cold_zone + 50)
-    frozen_zone = randint(gas_zone, gas_zone + 100)
-    '''
-    grammar = read_grammar("grammar/shitty_grammar.json")
-    assert grammar
-    gal_str = grammar.generate()
-    assert gal_str
-    assert len(gal_str) > 0
-    print("gal_str " + gal_str)
-
-    bit = gal_str.split(' ')
-    # skip bit[0] because that is star
-    hot_zone = randint(0, int(bit[1]))
-    bio_zone = randint(hot_zone, hot_zone + int(bit[2]))
-    cold_zone = randint(bio_zone, bio_zone + int(bit[3]))
-    gas_zone = randint(cold_zone, cold_zone + int(bit[4]))
-    frozen_zone = randint(gas_zone, gas_zone + int(bit[5]))
-
-    print(" zones: " + str([hot_zone,bio_zone,cold_zone,gas_zone,frozen_zone]))
-    '''
-    "_dynamic_tag": ["[star_type:#star_type<$star$>#]"],
-
-    num_planets = randint(2, 10)
-    p1 = 6 + randint(3, 10)
-    p2 = randint(p1+3, p1+7)
-    current_angle = 0
-    for i in range(1,num_planets):
-        current_angle = current_angle + 120
-        if i == 1:
-            planet_radius = p1 * self.system_scalar
-        elif i == 2:
-            planet_radius = p2 * self.system_scalar
-        else:
-            planet_radius = self.titus_bode(p1, p2, i) * self.system_scalar
-    '''
-    num_planets = randint(2, 10)
-    p1 = 8 + randint(1, 8)
-    p2 = p1 + 2 + randint(1, 5)
-    current_angle = 0
-    planet_array = []
-    for i in range(1, num_planets + 1):
-        current_angle = current_angle + 120
-        if i == 1:
-            planet_radius = p1
-        elif i == 2:
-            planet_radius = p2
-        else:
-            planet_radius = titus_bode(p1, p2, i)
-        if planet_radius < hot_zone:
-            planet = grammar.generate('planet<hot>')
-        elif planet_radius < bio_zone:
-            planet = grammar.generate('planet<bio>')
-        elif planet_radius < cold_zone:
-            planet = grammar.generate('planet<cold>')
-        elif planet_radius < gas_zone:
-            planet = grammar.generate('planet<gas>')
-        elif planet_radius < frozen_zone:
-            planet = grammar.generate('planet<frozen>')
-        else:
-            planet = 'frozen_belt'
-        planet_array.append({'radius': planet_radius, 'angle': current_angle, 'planet': planet})
-
-    print(f"PLANET {planet_array}")
-    assert False
+def test_function():
+    root = Rule('root', ["#child.capitalize#"])
+    child = Rule('child', ["ant"])
+    grammar = Grammar({'root':root, 'child':child})
+    output = grammar.generate()
+    assert "Ant" == output
+
+def test_function_args():
+    root = Rule('root', ["#child.remove(ant)#"])
+    child = Rule('child', ["pants"])
+    grammar = Grammar({'root':root, 'child':child})
+    output = grammar.generate()
+    assert "ps" == output
 
```

Comment:
Add tests for functions in $var_derefs$, \<tags\>, [execs], and test chaining by both dot and comma in each scope type.

Reply:
Lmao can't request changes since i am the author, wtf github.